File: app/javascript/channels/location_channel.js
import consumer from "channels/consumer";

consumer.subscriptions.create("LocationChannel", {
  connected() {
    console.log("Connected...");
  },

  disconnected() {
    console.log("Disconnected...");
    const event = new CustomEvent("channel-disconnected");
    window.dispatchEvent(event);
  },

  received(data) {
    console.log("Location updated...");
    const event = new CustomEvent("location-changed");
    window.dispatchEvent(event, data);
  },
});

File: app/channels/location_channel.rb
# frozen_string_literal: true

class LocationChannel < ApplicationCable::Channel
  def subscribed
    stream_for(location)
  end

  def receive(data)
    broadcast_to(location, { event: 'clientMoved', uuid:, coordinates: data['coordinates'] })
  end

  def unsubscribed
    client = location.clients.find_by(uuid:)
    if client
      client.destroy
      location.decrement(:client_count)
    end
    broadcast_to(location, { event: 'clientDisconnected', uuid: })
    Turbo::StreamsChannel
      .broadcast_prepend_to(
        location,
        partial: 'layouts/flash',
        locals: { flash: { notice: "#{client.name} disconnected" } },
        target: 'flash'
      )
  end

  private

  def channel_name
    "location_#{location.key}"
  end

  def location
    @location ||= Location.find_by!(key: params[:key])
  end
end
